### ISSUE TITLE
t2 cli can't find my tessels

### ISSUE THREAD

**Issue by Frijol** [tessel/t2-cli#390]
_From @Frijol on October 22, 2015 4:3_

Been having this issue where sometimes `t2` commands just can't find Tessels. Typical output:

```ssh
➜  t2 provision
INFO Looking for your Tessel...
WARN No Tessels Found.
```

_Copied from original issue: tessel/cli#188_

**Comment by Frijol** [comment]
Tried debug scripts, hit timeouts:

```sh
➜ DEBUG=discovery,discovery:* t2 list
INFO Searching for nearby Tessels...
  discovery Will scan for LAN devices +0ms
  discovery Will scan for USB devices +1ms
  discovery Timeout hit! Waiting for pending to finish... +5s
  discovery Done! Stopping. +2ms
WARN No Tessels Found.
➜  DEBUG=discovery,discovery:* t2 list -t 30
INFO Searching for nearby Tessels...
  discovery Will scan for LAN devices +0ms
  discovery Will scan for USB devices +1ms
  discovery Timeout hit! Waiting for pending to finish... +30s
  discovery Done! Stopping. +2ms
WARN No Tessels Found.
```

**Comment by Frijol** [comment]
It's showing up on serial, that's good – this autocompletes:

```sh
➜ ls /dev/tty.usbmodem1412
```

**Comment by Frijol** [comment]
_From @johnnyman727 on October 22, 2015 4:5_

@Frijol can you `dterm` in and run `ps`? We're checking to see if the usbdaemon is running.

**Comment by Frijol** [comment]
froze:

```sh
➜  alias v2="dterm /dev/tty.usbmodem1412"
➜  v2

dterm> ^C
^C^C^C^C^C^]^]^]

exit
^]^C
```

**Comment by Frijol** [comment]
Unplugged/replugged Tessel & opened new terminal window. Dtermed in

```sh
➜  ~  v2
0] random: nonblocking pool is initialized



BusyBox v1.23.2 (2015-06-23 05:35:53 UTC) built-in shell (ash)

Tessel 2  /  Built on OpenWrt
root@Tessel-02A3C66A13C7:/#
```

**Comment by Frijol** [comment]
Results of `ps` - what are we looking for?

```sh
root@Tessel-02A3C66A13C7:/# ps
  PID USER       VSZ STAT COMMAND
    1 root      1436 S    /sbin/procd
    2 root         0 SW   [kthreadd]
    3 root         0 SW   [ksoftirqd/0]
    4 root         0 SW   [kworker/0:0]
    5 root         0 SW<  [kworker/0:0H]
    6 root         0 SW   [kworker/u2:0]
    7 root         0 SW<  [khelper]
    8 root         0 SW   [kworker/u2:1]
   69 root         0 SW<  [writeback]
   71 root         0 SW<  [bioset]
   74 root         0 SW<  [kblockd]
   76 root         0 SW   [kswapd0]
   77 root         0 SW   [kworker/0:1]
   78 root         0 SW   [fsnotify_mark]
   80 root         0 SW   [spi32766]
  223 root         0 SW<  [deferwq]
  270 root         0 SW   [kworker/u2:2]
  282 root         0 SW   [kworker/0:2]
  306 root         0 SWN  [jffs2_gcd_mtd6]
  371 root       900 S    /sbin/ubusd
  409 root      1488 S    /bin/ash --login
  779 root         0 SW<  [ipv6_addrconf]
  865 root         0 SW<  [krfcommd]
  915 root         0 SW<  [cfg80211]
 1023 root      1056 S    /sbin/logd -S 16
 1032 root      1864 S    /sbin/rpcd
 1066 root      1572 S    /sbin/netifd
 1100 root      1192 S    /usr/sbin/odhcpd
 1138 root      1148 S    /usr/sbin/dropbear -F -P /var/run/dropbear.1.pid -s
 1147 root      1732 S    /usr/sbin/dbus-daemon --system
 1157 nobody    1324 S    /usr/sbin/mdnsd -debug
 1166 root       772 S    spid /dev/spidev32766.1 2 1 /var/run/tessel
 1209 root      1484 S    udhcpc -p /var/run/udhcpc-eth0.pid -s /lib/netifd/dh
 1271 root      1140 S    /usr/bin/mDNSResponder -b -f /etc/mDNSResponder.conf
 1281 root       904 S    dns-sd -R Tessel-02A3C66A13C7 _tessel._tcp local 22
 1291 root      3436 S    /usr/bin/bluetoothd -n
 1314 nobody     980 S    /usr/sbin/dnsmasq -C /var/etc/dnsmasq.conf -k -x /va
 1344 root      1484 S    /usr/sbin/ntpd -n -S /usr/sbin/ntpd-hotplug -p 0.ope
 1356 root       780 S    usbexecd /var/run/tessel/usb
 1361 root      1484 R    ps
```

**Comment by Frijol** [comment]
(also exited dterm, still experiencing the problem):

```sh
➜  ~  t2 list
INFO Searching for nearby Tessels...
WARN No Tessels Found.
```

**Comment by Frijol** [comment]
_From @johnnyman727 on October 22, 2015 4:13_

So we can see the usb daemon is running in this line:
```
1356 root       780 S    usbexecd /var/run/tessel/usb
```

To test if the usb daemon might be in a bad state, try running `/etc/init.d/usbexecd restart`. That will restart the usb daemon. Then try running `DEBUG=discovery,discovery:* t2 list` again (from outside dterm).

**Comment by Frijol** [comment]
ugh froze starting dterm again.

New terminal window. Tried starting dterm from there without rebooting Tessel. Froze again.

New terminal window. Rebooted Tessel. This time it worked - so something on Tessel's USB comms isn't closing right.

**Comment by johnnyman727** [comment]
Can you give https://github.com/tessel/t2-cli/pull/386 a try?

**Comment by Frijol** [comment]
Did the usbexecd restart within dterm, then ran the DEBUG t2 list again. It found my Tessel!

```sh
➜  ~  DEBUG=discovery,discovery:* t2 list
INFO Searching for nearby Tessels...
  discovery Will scan for LAN devices +0ms
  discovery Will scan for USB devices +2ms
  discovery:usb Device found. +5ms
  discovery Connection opened: +11ms undefined
  discovery Fetching name: +0ms undefined
  discovery:lan Device found: +453ms Hangy._tessel._tcp.local
  discovery:lan Device open error: +3s Hangy.local
  discovery Connection opened: +2ms Hangy.local
	Hangy	LAN	(USB connect and run `t2 provision` to authorize)
  discovery Timeout hit! Waiting for pending to finish... +2s
  discovery Done! Stopping. +3ms
```

Let me try to run some commands before I check out #386

**Comment by Frijol** [comment]
Welp, that was it. It found that usb Device (notably never printed it). Then on a second run of that command, we don't even get that:

```sh
➜  ~  DEBUG=discovery,discovery:* t2 list
INFO Searching for nearby Tessels...
  discovery Will scan for LAN devices +0ms
  discovery Will scan for USB devices +2ms
  discovery:lan Device found: +136ms Hangy._tessel._tcp.local
  discovery:lan Device open error: +2s Hangy.local
  discovery Connection opened: +4ms Hangy.local
	Hangy	LAN	(USB connect and run `t2 provision` to authorize)
  discovery Timeout hit! Waiting for pending to finish... +3s
  discovery Done! Stopping. +3ms
```

**Comment by Frijol** [comment]
Checking out #386 now

**Comment by Frijol** [comment]
On #386, looks the same:

```sh
➜  ~  DEBUG=discovery,discovery:* t2 list
INFO Searching for nearby Tessels...
  discovery Will scan for LAN devices +0ms
  discovery Will scan for USB devices +2ms
  discovery Timeout hit! Waiting for pending to finish... +5s
  discovery Done! Stopping. +2ms
WARN No Tessels Found.
```

**Comment by Frijol** [comment]
`/etc/init.d/usbexecd restart` is working, I just got myself on wifi & provisioned so now I'm not dependent on the USB comms... anything else I should try to help pinpoint the issue, thought?

**Comment by Frijol** [comment]
Managed an update (LEDs currently flashing a reboot) but got this angry readout:

```sh
➜  ~  t2 update
INFO Looking for your Tessel...
WARN -1 Tessel devices connected by USB may still be booting.
INFO Connected to Tessel-02A3FD15D69B over LAN
WARN Could not find firmware version on Tessel-02A3FD15D69B
WARN Forcefully updating...
INFO Updating Tessel-02A3FD15D69B to latest version (0.0.4)...
INFO Beginning update download. This could take a couple minutes..
  downloading [====================] 100% 0.0s remaining
INFO Download complete!
INFO Updating OpenWRT (1/2)
INFO Transferring image of size 19.92 MB
INFO Starting OpenWRT update. Please do not remove power from Tessel.
INFO Updating firwmare (2/2)
INFO Firmware update complete!
INFO Tessel is rebooting. Please wait about 30 seconds before using Tessel.
INFO Force updated Tessel-02A3FD15D69B to version 0.0.4
Unhandled rejection Error: LIBUSB_ERROR_NO_DEVICE
    at Error (native)
    at InEndpoint.Endpoint.stopPoll (/Users/frijol/t2-cli/node_modules/usb/usb.js:259:25)
    at USB.Connection._close (/Users/frijol/t2-cli/lib/usb_connection.js:181:13)
    at tryCatcher (/Users/frijol/t2-cli/node_modules/bluebird/js/main/util.js:26:23)
    at ret (eval at <anonymous> (/Users/frijol/t2-cli/node_modules/bluebird/js/main/promisify.js:163:12), <anonymous>:11:39)
    at /Users/frijol/t2-cli/lib/usb_connection.js:359:51
    at tryCatcher (/Users/frijol/t2-cli/node_modules/bluebird/js/main/util.js:26:23)
    at MappingPromiseArray._promiseFulfilled (/Users/frijol/t2-cli/node_modules/bluebird/js/main/map.js:56:38)
    at MappingPromiseArray.init [as _init$] (/Users/frijol/t2-cli/node_modules/bluebird/js/main/promise_array.js:92:18)
    at MappingPromiseArray.init (/Users/frijol/t2-cli/node_modules/bluebird/js/main/map.js:29:23)
    at Async._drainQueue (/Users/frijol/t2-cli/node_modules/bluebird/js/main/async.js:128:12)
    at Async._drainQueues (/Users/frijol/t2-cli/node_modules/bluebird/js/main/async.js:133:10)
    at Immediate.Async.drainQueues [as _onImmediate] (/Users/frijol/t2-cli/node_modules/bluebird/js/main/async.js:15:14)
    at processImmediate [as _immediateCallback] (timers.js:368:17)
```

**Comment by Frijol** [comment]
Interestingly, it can still find it over USB.

```sh
➜  ~  t2 list
INFO Searching for nearby Tessels...
	Tessel-02A3FD15D69B	USB
```

...and even weirder, now that I think about it, is that it didn't reconnect to Wifi automatically.

**Comment by johnnyman727** [comment]
Yes, I've seen that issue where *sometimes* an OpenWRT update will wipe the wireless credentials. Just made an open issue: https://github.com/tessel/t2-cli/issues/391

**Comment by Frijol** [comment]
Trying another T2 that needed updating, this time not able to fix it

```sh
➜  ~  t2 provision
INFO Looking for your Tessel...
WARN 1 Tessel device connected by USB may still be booting.
WARN No Tessels Found.
➜  ~  v2

root@Tessel-02A3C66A13C7:/# /etc/init.d/usbexecd restart
root@Tessel-02A3C66A13C7:/#
dterm> ^C

➜  ~  t2 provision
INFO Looking for your Tessel...
WARN 1 Tessel device connected by USB may still be booting.
WARN No Tessels Found.
```

**Comment by Frijol** [comment]
Using a different Tessel (still on #386) getting this issue:

```sh
➜  ~  t2 wifi
INFO Looking for your Tessel...
WARN 1 Tessel device connected by USB may still be booting.
WARN No Authorized Tessels Found.
events.js:141
      throw er; // Unhandled 'error' event
            ^
Error: Connection was already closed...
    at USB.Connection._write (/Users/frijol/t2-cli/lib/usb_connection.js:70:14)
    at doWrite (_stream_writable.js:291:12)
    at clearBuffer (_stream_writable.js:401:7)
    at onwrite (_stream_writable.js:336:7)
    at OutEndpoint.WritableState.onwrite (_stream_writable.js:88:5)
    at Transfer.callback (/Users/frijol/t2-cli/node_modules/usb/usb.js:343:14)
```

**Comment by jbergstroem** [comment]
Not sure if dup or no but I'm having similar issues:
```
~/Work/tessel $ t2 version
INFO Looking for your Tessel...
INFO Connected to Tessel-02A30D3B52A3 over USB
events.js:85
      throw er; // Unhandled 'error' event
            ^
Error: Error upon remote process exit: 1
    at USBProcess.<anonymous> (/Users/jbergstroem/.nvm/versions/node/v0.12.7/lib/node_modules/t2-cli/lib/usb/usb_process.js:49:26)
    at USBProcess.g (events.js:199:16)
    at USBProcess.emit (events.js:107:17)
    at Parser.<anonymous> (/Users/jbergstroem/.nvm/versions/node/v0.12.7/lib/node_modules/t2-cli/lib/usb/usb_daemon.js:121:14)
    at Parser.emit (events.js:107:17)
    at Parser._write (/Users/jbergstroem/.nvm/versions/node/v0.12.7/lib/node_modules/t2-cli/node_modules/usb-daemon-parser/index.js:152:16)
    at doWrite (_stream_writable.js:301:12)
    at writeOrBuffer (_stream_writable.js:288:5)
    at Parser.Writable.write (_stream_writable.js:217:11)
    at ondata (_stream_readable.js:540:20)
    at emit (events.js:107:17)
    at readableAddChunk (_stream_readable.js:163:16)
    at Readable.push (_stream_readable.js:126:10)
    at InEndpoint.emit (events.js:107:17)
    at Transfer.transferDone (/Users/jbergstroem/.nvm/versions/node/v0.12.7/lib/node_modules/t2-cli/node_modules/usb/usb.js:295:9)

~/Work/tessel $ system_profiler SPUSBDataType
USB:

    USB 3.0 Bus:

      Host Controller Driver: AppleUSBXHCIWPT
      PCI Device ID: 0x9cb1 
      PCI Revision ID: 0x0003 
      PCI Vendor ID: 0x8086 

        Bluetooth USB Host Controller:
        <snip for relevance>

        Tessel 2:

          Product ID: 0x7551
          Vendor ID: 0x1209
          Version: 1.10
          Serial Number: <changed if sensitive for whatever reason>
          Speed: Up to 12 Mb/sec
          Manufacturer: Technical Machine
          Location ID: 0x14100000 / 11
          Current Available (mA): 1000
          Current Required (mA): 500
          Extra Operating Current (mA): 0
```

..after the failed connection attempt it cannot be identified.

**Comment by jbergstroem** [comment]
FWIW, updating to `0.0.5` seems to have fixed above connectivity issues for me.

**Comment by Frijol** [comment]
You're getting that every time? That's usually what it looks like if you disconnect USB mid-connection

**Comment by johnnyman727** [comment]
@jbergstroem could you try rebooting your T2 and [trying out these debugging steps](https://github.com/tessel/t2-docs/blob/jon-usb-debug/debugging/usb.md)? I'd love to know 1. How many USB Daemon processes you see and 2. if restarting them before running the `t2 version` command help.

**Comment by jbergstroem** [comment]
@johnnyman727 restarting always helped. I've updated the firmware and it seems to a smooth ride from here on. Will have a closer look tomorrow.

**Comment by johnnyman727** [comment]
This should be fixed with #482 and #486.
